daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala

```diff
@@ -0,0 +1,173 @@
+      templateId: Ref.Identifier,
+      coid: Value.AbsoluteContractId,
+  ): speedy.Command.Fetch =
+    speedy.Command.Fetch(templateId, SValue.SContractId(coid))
+
+  @throws[PreprocessorException]
+  def unsafePreprocessExercise(
+      templateId: Ref.Identifier,
+      contractId: Value.ContractId,
+      choiceId: Ref.ChoiceName,
+      argument: Value[Value.AbsoluteContractId],
+  ): speedy.Command.Exercise = {
+    val template = unsafeGetTemplate(templateId)
+    val choiceArgType = unsafeGetChoiceArgType(templateId, template, choiceId)
+    val arg = valueTranslator.unsafeTranslateValue(choiceArgType, argument)
+    speedy.Command.Exercise(templateId, SValue.SContractId(contractId), choiceId, arg)
+  }
+
+  @throws[PreprocessorException]
+  def unsafePreprocessExerciseByKey(
+      templateId: Ref.Identifier,
+      contractKey: Value[Value.AbsoluteContractId],
+      choiceId: Ref.ChoiceName,
+      argument: Value[Value.AbsoluteContractId],
+  ): speedy.Command.ExerciseByKey = {
+    val template = unsafeGetTemplate(templateId)
+    val choiceArgType = unsafeGetChoiceArgType(templateId, template, choiceId)
+    val ckTtype = unsafeGetContractKeyType(templateId, template)
+    val arg = valueTranslator.unsafeTranslateValue(choiceArgType, argument)
+    val key = valueTranslator.unsafeTranslateValue(ckTtype, contractKey)
+    speedy.Command.ExerciseByKey(templateId, key, choiceId, arg)
+  }
+
+  @throws[PreprocessorException]
+  private[preprocessing] def unsafePreprocessCreateAndExercise(
+      templateId: Ref.ValueRef,
+      createArgument: Value[Value.AbsoluteContractId],
+      choiceId: Ref.ChoiceName,
+      choiceArgument: Value[Value.AbsoluteContractId],
+  ): speedy.Command.CreateAndExercise = {
+    val createArg =
+      valueTranslator.unsafeTranslateValue(Ast.TTyCon(templateId), createArgument)
+    val template = unsafeGetTemplate(templateId)
+    val choiceArgType = unsafeGetChoiceArgType(templateId, template, choiceId)
+    val choiceArg =
+      valueTranslator.unsafeTranslateValue(choiceArgType, choiceArgument)
+    speedy.Command
+      .CreateAndExercise(templateId, createArg, choiceId, choiceArg)
+  }
+
+  @throws[PreprocessorException]
+  private[preprocessing] def unsafePreprocessLookupByKey(
+      templateId: Ref.ValueRef,
+      contractKey: Value[Nothing],
+  ): speedy.Command.LookupByKey = {
+    val template = unsafeGetTemplate(templateId)
+    val ckTtype = unsafeGetContractKeyType(templateId, template)
+    val key = valueTranslator.unsafeTranslateValue(ckTtype, contractKey)
+    speedy.Command.LookupByKey(templateId, key)
+  }
+
+  @throws[PreprocessorException]
+  def unsafePreprocessCommands(
+      cmds: ImmArray[command.Command],
+  ): ImmArray[speedy.Command] = {
+    // before, we had
+    //
+    // ```
+    // Result.sequence(ImmArray(cmds.commands).map(preprocessCommand))
+    // ```
+    //
+    // however that is bad, because it'll generate a `NeedPackage` for each command,
```

Comment:
This would only happen on the first set of command that refer to a certain package, right?

Reply:
Yes. 
Anyway, this comment is legacy in the sens we want to get rid of the "monadic" nesting.